src/bridge/events.rs

```diff
@@ -159,494 +160,458 @@ fn unpack_color(packed_color: u64) -> Color4f {
+    extract_values!([content], msg_showmode_arguments);
+
+    Ok(RedrawEvent::MessageShowMode {
+        content: parse_styled_content(content)?,
+    })
 }
 
-fn parse_msg_showcmd(msg_showcmd_arguments: &[Value]) -> Result<RedrawEvent> {
-    if let [content] = msg_showcmd_arguments {
-        Ok(RedrawEvent::MessageShowCommand {
-            content: parse_styled_content(content)?,
-        })
-    } else {
-        Err(EventParseError::InvalidEventFormat)
-    }
+fn parse_msg_showcmd(msg_showcmd_arguments: Vec<Value>) -> Result<RedrawEvent> {
+    extract_values!([content], msg_showcmd_arguments);
+
+    Ok(RedrawEvent::MessageShowCommand {
+        content: parse_styled_content(content)?,
+    })
 }
 
-fn parse_msg_ruler(msg_ruler_arguments: &[Value]) -> Result<RedrawEvent> {
-    if let [content] = msg_ruler_arguments {
-        Ok(RedrawEvent::MessageRuler {
-            content: parse_styled_content(content)?,
-        })
-    } else {
-        Err(EventParseError::InvalidEventFormat)
-    }
+fn parse_msg_ruler(msg_ruler_arguments: Vec<Value>) -> Result<RedrawEvent> {
+    extract_values!([content], msg_ruler_arguments);
+
+    Ok(RedrawEvent::MessageRuler {
+        content: parse_styled_content(content)?,
+    })
 }
 
-fn parse_msg_history_entry(entry: &Value) -> Result<(MessageKind, StyledContent)> {
-    if let [kind, content] = parse_array(entry)?{
-        Ok((
-            MessageKind::parse(parse_string(kind)?),
-            parse_styled_content(content)?
-        ))
-    } else {
-        Err(EventParseError::InvalidEventFormat)
-    }
+fn parse_msg_history_entry(entry: Value) -> Result<(MessageKind, StyledContent)> {
+    extract_values!([kind, content], parse_array(entry)?);
+
+    Ok((
+        MessageKind::parse(&parse_string(kind)?),
+        parse_styled_content(content)?
+    ))
 }
 
-fn parse_msg_history_show(msg_history_show_arguments: &[Value]) -> Result<RedrawEvent> {
-    if let [entries] = msg_history_show_arguments {
-        Ok(RedrawEvent::MessageHistoryShow {
-            entries: parse_array(entries)?
-                .iter()
-                .map(parse_msg_history_entry)
-                .collect::<Result<_>>()?
-        })
-    } else {
-        Err(EventParseError::InvalidEventFormat)
-    }
+fn parse_msg_history_show(msg_history_show_arguments: Vec<Value>) -> Result<RedrawEvent> {
+    extract_values!([entries], msg_history_show_arguments);
+
+    Ok(RedrawEvent::MessageHistoryShow {
+        entries: parse_array(entries)?
+            .into_iter()
+            .map(parse_msg_history_entry)
+            .collect::<Result<_>>()?
+    })
 }
 
-pub fn parse_redraw_event(event_value: &Value) -> Result<Vec<RedrawEvent>> {
-    let event_contents = parse_array(event_value)?;
-    let name_value = event_contents.get(0).ok_or(EventParseError::InvalidEventFormat)?;
+pub fn parse_redraw_event(event_value: Value) -> Result<Vec<RedrawEvent>> {
+    let mut event_contents = parse_array(event_value)?.into_iter();
```

Comment:
This is a micro optimization which I think obscures the original intention (take the first for the name and the rest for the arguments) for little observable performance gain. Event parsing doesn't show up on the profiling of the app as an appreciable performance concern, so I would rather leave it as it makes the code more readable.

I realize that the previous code took it as a reference, so I'm not sure what the exact right strategy is here, but if at all possible I would like to preserve the initial reading where it is obvious that the name is the first and the arguments are the rest. This could be achieved just with properly named variables... Not sure.

That said, this is relatively nit picky so take it or leave it.

Reply:
Actually, I didn't do it as an optimization 😅, was due to the fact that we need ownership of the values.

> think obscures the original intention (take the first for the name and the rest for the arguments)

Actually, this code still expresses this intention, as we take the first element of the iterator (`event_contents.next()`) and use the rest of it for the arguments (`events = event_contents`).

> This could be achieved just with properly named variables... Not sure.

Maybe, I will try to find better names.